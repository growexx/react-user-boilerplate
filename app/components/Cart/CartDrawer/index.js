import {
  Affix,
  Avatar,
  Button,
  Col,
  Divider,
  Drawer,
  List,
  message,
  Row,
  Typography,
} from 'antd';
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { DollarCircleOutlined } from '@ant-design/icons';

const { Text } = Typography;
const CartDrawer = ({ visible, setVisible }) => {
  const [productsData, setProductsData] = useState(window.product || []);
  useEffect(() => {
    setProductsData(window.product || []);
  }, [window.product]);
  useEffect(() => {
    localStorage.setItem('products', JSON.stringify(productsData));
  }, [productsData]);
  useEffect(() => {
    window.addEventListener('storage', () => {
      setProductsData(JSON.parse(localStorage.getItem('products')) || []);
    });
    return () => {
      window.removeEventListener('storage', window);
    };
  }, []);
  const onDeleteClick = id => {
    const filterProductsData = productsData.filter(
      product => product.id !== id,
    );
    window.setCount(filterProductsData.length);
    setProductsData(filterProductsData);
    message.success('Product Deleted Successfully from cart');
  };

  const onCloseHandler = () => {
    setVisible(false);
  };

  const total = productsData
    .map(product => product.price)
    .reduce((a, b) => a + b, 0);
  return (
    <Drawer
      title="Product Cart"
      placement="right"
      closable
      onClose={onCloseHandler}
      visible={visible}
      width="320"
      data-testid="drawer"
    >
      <List
        itemLayout="vertical"
        dataSource={productsData}
        renderItem={product => (
          <List.Item key={product.id}>
            <List.Item.Meta
              avatar={<Avatar src={product.imageUrl} />}
              title={
                <div className="u-d-flex u-align-items-center u-justify-content-between">
                  {product.title}
                  <div className="u-d-flex u-align-items-center">
                    <DollarCircleOutlined className="u-mr-1" />
                    {product.price}
                  </div>
                </div>
              }
              description={product.description}
            />
            <Row>
              <Col span={4} />
              <Col span={20}>
                <div className="u-d-flex u-justify-content-between">
                  <Text type="secondary">Product Qty:</Text>
                  <Text>{product.qty}</Text>
                </div>
              </Col>
              <Col span={4} className="u-mt-2" />
              <Col span={20} className="u-mt-2">
                <Button
                  size="small"
                  onClick={() => onDeleteClick(product.id)}
                  data-testid="product-delete"
                >
                  Delete
                </Button>
              </Col>
            </Row>
          </List.Item>
        )}
      />
      <Affix offsetBottom={0}>
        <div>
          <Divider />

          <div className="u-d-flex u-justify-content-between">
            <Text>Total Payable Amount:</Text>
            <Text className="u-d-flex u-align-items-center">
              <DollarCircleOutlined className="u-mr-2" />
              {total}
            </Text>
          </div>
        </div>
      </Affix>
    </Drawer>
  );
};

CartDrawer.propTypes = {
  visible: PropTypes.bool,
  setVisible: PropTypes.func,
};
export default CartDrawer;
