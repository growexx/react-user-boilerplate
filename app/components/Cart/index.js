/**
 * Cart/index.js
 *
 * This is the Cart Component file.
 */
import React, { useEffect, useState } from 'react';
import { ShoppingCartOutlined } from '@ant-design/icons';
import { Badge, Button } from 'antd';
import CartDrawer from './CartDrawer';

const Cart = () => {
  const productCount = JSON.parse(localStorage.products || '[]').length;
  const [visible, setVisible] = useState(false);
  const [count, setCount] = useState(productCount);
  const onClickHandler = () => {
    setVisible(true);
  };

  useEffect(() => {
    window.setCount = setCount;
    window.addEventListener('storage', () => {
      setCount(JSON.parse(localStorage.getItem('products')).length || []);
    });
    return () => {
      window.removeEventListener('storage', window);
    };
  }, []);

  useEffect(() => {
    setCount((window.product || []).length);
  }, [window.product]);
  return (
    <>
      <div className="u-mr-1 u-d-inline-block">
        <Button
          onClick={onClickHandler}
          type="text"
          data-testid="badge-Cart"
          className="btn-hover-none p-4"
        >
          <Badge count={count} size="small">
            <ShoppingCartOutlined className="u-font-size-xlg" />
          </Badge>
        </Button>
      </div>
      <div data-testid="badge-cart-drawer" data-visible={visible}>
        <CartDrawer visible={visible} setVisible={setVisible} />
      </div>
    </>
  );
};
export default Cart;
