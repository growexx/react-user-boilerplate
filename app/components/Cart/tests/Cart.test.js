import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { ConnectedRouter } from 'connected-react-router/immutable';
import { createMemoryHistory } from 'history';
import products from '../../../examples/Products/stub/product.json';

import Cart from '../index';
import configureStore from '../../../configureStore';

const dummyData = products.products.slice(0, 2);

describe('<Cart />', () => {
  const history = createMemoryHistory();
  const store = configureStore({}, history);

  it('should render a div', () => {
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Cart />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('<Cart />', () => {
  test('display should update Carts', () => {
    window.product = dummyData;
    window.localStorage = {};
    window.localStorage.setItem = (key, value) => {
      window.localStorage[key] = value;
    };
    window.localStorage.getItem = key => window.localStorage[key];
    const { getByTestId } = render(<Cart />);
    window.localStorage.setItem('products', JSON.stringify(dummyData));
    window.dispatchEvent(
      new StorageEvent('storage', {
        key: 'test_key',
        newValue: 'test_value',
      }),
    );
    fireEvent.click(getByTestId('badge-Cart'));

    expect(getByTestId('badge-cart-drawer')).toBeTruthy();
  });
});
