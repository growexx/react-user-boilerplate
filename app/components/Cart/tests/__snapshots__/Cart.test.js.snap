// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> should render a div 1`] = `null`;
